src/libostree/ostree-sign-ed25519.c

```diff
@@ -0,0 +1,714 @@
+    {
+      const gchar *pk_ascii = g_variant_get_string (public_key, NULL);
+      key = g_base64_decode (pk_ascii, &n_elements);
+    }
+  else if (g_variant_is_of_type (public_key, G_VARIANT_TYPE_BYTESTRING))
+    {
+      key = (gpointer) g_variant_get_fixed_array (public_key, &n_elements, sizeof(guchar));
+    }
+  else
+    {
+      return glnx_throw (error, "Unknown ed25519 public key type");
+    }
+
+  g_autofree char *hex = g_malloc0 (crypto_sign_PUBLICKEYBYTES*2 + 1);
+  g_debug ("Read ed25519 public key = %s", sodium_bin2hex (hex, crypto_sign_PUBLICKEYBYTES*2+1, key, n_elements));
+
+  if (n_elements != crypto_sign_PUBLICKEYBYTES)
+    return glnx_throw (error, "Incorrect ed25519 public key");
+
+  if (g_list_find_custom (sign->public_keys, key, _compare_ed25519_keys) == NULL)
+    {
+      gpointer newkey = g_memdup (key, n_elements);
+      sign->public_keys = g_list_prepend (sign->public_keys, newkey);
+    }
+
+#endif /* HAVE_LIBSODIUM */
+  return TRUE;
+}
+
+#ifdef HAVE_LIBSODIUM
+/* Add revoked public key */
+static gboolean
+_ed25519_add_revoked (OstreeSign *self,
+                      GVariant *revoked_key,
+                      GError **error)
+{
+  g_debug ("%s enter", __FUNCTION__);
+  g_return_val_if_fail (OSTREE_IS_SIGN (self), FALSE);
+
+  if (!g_variant_is_of_type (revoked_key, G_VARIANT_TYPE_STRING))
+    return glnx_throw (error, "Unknown ed25519 revoked key type");
+
+  OstreeSignEd25519 *sign = _ostree_sign_ed25519_get_instance_private(OSTREE_SIGN_ED25519(self));
+
+  const gchar *rk_ascii = g_variant_get_string (revoked_key, NULL);
+  gsize n_elements = 0;
+  gpointer key = g_base64_decode (rk_ascii, &n_elements);
+
+  g_autofree char * hex = g_malloc0 (crypto_sign_PUBLICKEYBYTES*2 + 1);
+  g_debug ("Read ed25519 revoked key = %s", sodium_bin2hex (hex, crypto_sign_PUBLICKEYBYTES*2+1, key, n_elements));
+
+  if (n_elements != crypto_sign_PUBLICKEYBYTES)
+    {
+      return glnx_throw (error, "Incorrect ed25519 revoked key");
+    }
+
+  if (g_list_find_custom (sign->revoked_keys, key, _compare_ed25519_keys) == NULL)
+    {
+      gpointer newkey = g_memdup (key, n_elements);
+      sign->revoked_keys = g_list_prepend (sign->revoked_keys, newkey);
+    }
+
+  return TRUE;
+}
+#endif /* HAVE_LIBSODIUM */
+
+
+static gboolean
+_load_pk_from_stream (OstreeSign *self,
+                      GDataInputStream *key_data_in,
+                      gboolean trusted,
+                      GError **error)
+{
+  g_debug ("%s enter", __FUNCTION__);
+  g_return_val_if_fail (key_data_in, FALSE);
+#ifdef HAVE_LIBSODIUM
+  gboolean ret = FALSE;
+
+  /* Use simple file format with just a list of base64 public keys per line */
+  while (TRUE)
+    {
+      gsize len = 0;
+      g_autofree char *line = g_data_input_stream_read_line (key_data_in, &len, NULL, error);
+      g_autoptr (GVariant) pk = NULL;
+      gboolean added = FALSE;
+
+      if (*error != NULL)
+        goto err;
```

Comment:
I'm finding it hard to get over the error handling style - I know e.g. the Linux kernel makes heavy use of `goto` but again in all of our new code we try to avoid it.

Actually this case needs to be:

```
if (!line)
  return FALSE;
```

Because it's supported again to pass `NULL` for `GError`.